File: services/recommendation-service/src/services/tmdbClient.js


const axios = require("axios")

const BASE_URL = "https://api.themoviedb.org/3"

const tmdbApi = axios.create({
  baseURL: BASE_URL,
  timeout: 10000,
  headers: { Accept: "application/json" }
})
const _cache = new Map()
const CACHE_TTL = 15 * 60 * 1000

const cacheGet = (key) => {
  const entry = _cache.get(key)
  if (!entry) return null
  if (Date.now() - entry.ts > CACHE_TTL) {
    _cache.delete(key)
    return null
  }
  return entry.data
}

const cacheSet = (key, data) => {
  _cache.set(key, { data, ts: Date.now() })
}

const fetchWithRetry = async (url, params, retries = 3) => {
  for (let i = 0; i < retries; i++) {
    try {
      const response = await tmdbApi.get(url, { params })
      return response
    } catch (error) {
      if (i === retries - 1) throw error
      await new Promise(resolve => setTimeout(resolve, 1000 * (i + 1)))
    }
  }
}

const fetchPopularMovies = async (pages = 5) => {
  const cacheKey = `popular_bulk_${pages}`
  const cached = cacheGet(cacheKey)
  if (cached) return cached

  const allMovies = []
  for (let page = 1; page <= pages; page++) {
    try {
      const response = await fetchWithRetry("/movie/popular", {
        api_key: process.env.TMDB_API_KEY,
        page
      })
      allMovies.push(...response.data.results)
      if (page < pages) {
        await new Promise(resolve => setTimeout(resolve, 200))
      }
    } catch (error) {
      console.error(`[tmdbClient] Failed to fetch popular page ${page}:`, error.message)
    }
  }

  cacheSet(cacheKey, allMovies)
  return allMovies
}

const fetchMovieDetails = async (movieId) => {
  const cacheKey = `movie_${movieId}`
  const cached = cacheGet(cacheKey)
  if (cached) return cached

  const response = await fetchWithRetry(`/movie/${movieId}`, {
    api_key: process.env.TMDB_API_KEY
  })

  cacheSet(cacheKey, response.data)
  return response.data
}

const fetchMoviesByIds = async (movieIds) => {
  const movieMap = new Map()
  const uncachedIds = movieIds.filter(id => {
    const cached = cacheGet(`movie_${id}`)
    if (cached) {
      movieMap.set(id, cached)
      return false
    }
    return true
  })

  const BATCH_SIZE = 10
  for (let i = 0; i < uncachedIds.length; i += BATCH_SIZE) {
    const batch = uncachedIds.slice(i, i + BATCH_SIZE)
    
    const results = await Promise.all(
      batch.map(async (id) => {
        try {
          const movie = await fetchMovieDetails(id)
          return { id, movie }
        } catch (error) {
          console.warn(`[tmdbClient] Failed to fetch movie ${id}:`, error.message)
          return { id, movie: null }
        }
      })
    )

    for (const { id, movie } of results) {
      if (movie) movieMap.set(id, movie)
    }

    if (i + BATCH_SIZE < uncachedIds.length) {
      await new Promise(resolve => setTimeout(resolve, 300))
    }
  }

  return movieMap
}

const discoverByGenre = async (genreId, page = 1) => {
  const cacheKey = `discover_genre_${genreId}_${page}`
  const cached = cacheGet(cacheKey)
  if (cached) return cached

  const response = await fetchWithRetry("/discover/movie", {
    api_key: process.env.TMDB_API_KEY,
    with_genres: genreId,
    sort_by: "popularity.desc",
    "vote_count.gte": 50,
    page
  })

  cacheSet(cacheKey, response.data.results)
  return response.data.results
}

const fetchTrendingMovies = async (limit = 40) => {
  const cacheKey = `trending_${limit}`
  const cached = cacheGet(cacheKey)
  if (cached) return cached

  const results = []
  const pages = Math.ceil(limit / 20)

  for (let page = 1; page <= pages; page++) {
    try {
      const response = await fetchWithRetry("/trending/movie/week", {
        api_key: process.env.TMDB_API_KEY,
        page
      })
      results.push(...response.data.results)
    } catch (error) {
      console.error(`[tmdbClient] Failed to fetch trending page ${page}:`, error.message)
    }
  }

  const sliced = results.slice(0, limit)
  cacheSet(cacheKey, sliced)
  return sliced
}

module.exports = {
  fetchPopularMovies,
  fetchMovieDetails,
  fetchMoviesByIds,
  discoverByGenre,
  fetchTrendingMovies
}

File: services/ml-service/tmdb_client.py


import os
import time
import requests
from requests.adapters import HTTPAdapter
from urllib3.util.retry import Retry

TMDB_API_KEY = os.getenv("TMDB_API_KEY", "")
BASE_URL = "https://api.themoviedb.org/3"
CACHE_TTL = 900
_cache = {}


def _cache_get(key):
    entry = _cache.get(key)
    if entry is None:
        return None
    if time.time() - entry["ts"] > CACHE_TTL:
        del _cache[key]
        return None
    return entry["data"]


def _cache_set(key, data):
    _cache[key] = {"data": data, "ts": time.time()}

_session = requests.Session()
_retry_strategy = Retry(
    total=5,
    backoff_factor=1,
    status_forcelist=[429, 500, 502, 503, 504],
    allowed_methods=["GET"]
)
_adapter = HTTPAdapter(max_retries=_retry_strategy)
_session.mount("http://", _adapter)
_session.mount("https://", _adapter)

def _tmdb_get(path, params=None):
    if params is None:
        params = {}
    params["api_key"] = TMDB_API_KEY

    try:
        resp = _session.get(f"{BASE_URL}{path}", params=params, timeout=30)
        resp.raise_for_status()
        return resp.json()
    except Exception as e:
        print(f"[tmdb_client] ERROR for {path}: {e}")
        raise


def fetch_popular_movies(pages=5):
    cache_key = f"popular_{pages}"
    cached = _cache_get(cache_key)
    if cached is not None:
        return cached

    all_movies = []
    for page in range(1, pages + 1):
        try:
            data = _tmdb_get("/movie/popular", {"page": page})
            for m in data.get("results", []):
                all_movies.append({
                    "movie_id": m["id"],
                    "title": m.get("title", ""),
                    "overview": m.get("overview", ""),
                    "poster_path": m.get("poster_path", ""),
                    "vote_average": m.get("vote_average", 0),
                    "vote_count": m.get("vote_count", 0),
                    "popularity": m.get("popularity", 0),
                    "release_date": m.get("release_date", ""),
                    "language": m.get("original_language", "en"),
                    "genre_ids": m.get("genre_ids", [])
                })
            if page < pages:
                time.sleep(0.5)
        except Exception as e:
            print(f"[tmdb_client] Failed to fetch popular page {page}: {e}")

    _cache_set(cache_key, all_movies)
    return all_movies


def fetch_trending_movies(limit=40):
    cache_key = f"trending_{limit}"
    cached = _cache_get(cache_key)
    if cached is not None:
        return cached

    results = []
    pages = (limit // 20) + 1
    for page in range(1, pages + 1):
        try:
            data = _tmdb_get("/trending/movie/week", {"page": page})
            for m in data.get("results", []):
                results.append({
                    "movie_id": m["id"],
                    "title": m.get("title", ""),
                    "overview": m.get("overview", ""),
                    "poster_path": m.get("poster_path", ""),
                    "vote_average": m.get("vote_average", 0),
                    "vote_count": m.get("vote_count", 0),
                    "popularity": m.get("popularity", 0),
                    "release_date": m.get("release_date", ""),
                    "language": m.get("original_language", "en"),
                    "genre_ids": m.get("genre_ids", [])
                })
        except Exception as e:
            print(f"[tmdb_client] Failed to fetch trending page {page}: {e}")

    sliced = results[:limit]
    _cache_set(cache_key, sliced)
    return sliced


def fetch_discover_movies(language="en", pages=2):
    cache_key = f"discover_lang_{language}_{pages}"
    cached = _cache_get(cache_key)
    if cached is not None:
        return cached

    results = []
    for page in range(1, pages + 1):
        try:
            data = _tmdb_get("/discover/movie", {
                "with_original_language": language,
                "sort_by": "popularity.desc",
                "page": page
            })
            for m in data.get("results", []):
                results.append({
                    "movie_id": m["id"],
                    "title": m.get("title", ""),
                    "overview": m.get("overview", ""),
                    "poster_path": m.get("poster_path", ""),
                    "vote_average": m.get("vote_average", 0),
                    "vote_count": m.get("vote_count", 0),
                    "popularity": m.get("popularity", 0),
                    "release_date": m.get("release_date", ""),
                    "language": m.get("original_language", "en"),
                    "genre_ids": m.get("genre_ids", [])
                })
        except Exception as e:
            print(f"[tmdb_client] Failed to fetch discover page {page}: {e}")

    _cache_set(cache_key, results)
    return results


def fetch_genre_list():
    cached = _cache_get("genre_list")
    if cached is not None:
        return cached

    data = _tmdb_get("/genre/movie/list")
    genres = {g["id"]: g["name"] for g in data.get("genres", [])}
    _cache_set("genre_list", genres)
    return genres


def fetch_movie_details(movie_id):
    cache_key = f"movie_details_{movie_id}"
    cached = _cache_get(cache_key)
    if cached is not None:
        return cached

    try:
        m = _tmdb_get(f"/movie/{movie_id}")
        movie = {
            "movie_id": m["id"],
            "title": m.get("title", ""),
            "overview": m.get("overview", ""),
            "poster_path": m.get("poster_path", ""),
            "vote_average": m.get("vote_average", 0),
            "vote_count": m.get("vote_count", 0),
            "popularity": m.get("popularity", 0),
            "release_date": m.get("release_date", ""),
            "language": m.get("original_language", "en"),
            "genre_ids": [g["id"] for g in m.get("genres", [])]
        }
        _cache_set(cache_key, movie)
        return movie
    except Exception as e:
        print(f"[tmdb_client] Failed to fetch details for movie {movie_id}: {e}")
        return None
